src/libertem/io/dataset/mib.py

```diff
@@ -1230,6 +1270,29 @@ def get_base_shape(self, roi: Optional[np.ndarray]) -> Tuple[int, ...]:
         assert self.meta is not None and base_shape[-1] == self.meta.shape[-1]
         return base_shape
 
+    def get_num_partitions(self) -> int:
+        # Try to limit the partition to some fraction of the max_open files limit
+        limit_n_files = float('inf')
+        if platform.system() != "Windows":
+            import resource
+            # Windows has no problem opening huge numbers of files
+            limit_n_files, _ = resource.getrlimit(resource.RLIMIT_NOFILE)
```

Comment:
Maybe we can subtract the already open files here? Should be accessible via `getrusage(resource.RUSAGE_SELF)`

Additional comment: on Linux, one of the main reasons to limit the number of open file descriptors is the `select` syscall, which we should not be using anyways. The hard limit of file descriptors is a lot higher than 1024, for example on my system:

![image](https://github.com/LiberTEM/LiberTEM/assets/5778/b48ff505-95c7-47ec-a5cd-3a6b6782c1b5)

1048576 files should be enough for everyone™ (another common value seems to be 262144 on CentOS machines; on older versions though the hard limit is often indeed smaller, I've seen something like 4096). So, what we could do, is increase the soft limit to be equal to the hard limit. This should be done "globally" in our worker processes, not as a per-dataset-type fix. We can include something like this in the worker setup code (dask, pipelined):

```python
_, hard_lim = resource.getrlimit(resource.RLIMIT_NOFILE)
resource.setrlimit(resource.RLIMIT_NOFILE, (hard_lim, hard_lim))
```

(aside: this is also something we should include in our testing setup, where I sometimes manually set the `ulimit` before running `pytest`)

For threaded workers, it becomes a bit harder, as we don't "own" the Python process itself, and might cause issues if we modify the process-wide limit. Do you have an opinion on this? Or should we assume that `select` usage is rare, and work on an opt-out basis?

If setting the rlimits is not enough (for systems with a low hard limit), instead of a per-dataset-type fix here, we should fix the backends to limit the number of concurrently open files instead - then you also don't need a separate fix for the raw `DataSet`.

Reply:
> So, what we could do, is increase the soft limit to be equal to the hard limit. 

I thought about this, but wasn't sure of what consequences it might create. If you're happy with it as an approach then I'll implement that instead.

> For threaded workers, it becomes a bit harder, as we don't "own" the Python process itself, and might cause issues if we modify the process-wide limit. Do you have an opinion on this?

I think that using processes is probably 90%+ of real usage so we're making an improvement. I don't know enough about the implications of a higher limit for the main Python process to have a strong opinion! I'll take a look at the implementation and see if we can even implement something global with the currently accessible info (I'm not sure if a dataset can know enough about the execution context at the moment that we call `get_partitions` ? Especially if we're running on an external cluster or on multiple-nodes etc etc).

Reply:
The way I wrote the test this assumes the test files are in `TESTDATA_BASE_PATH`. It takes about 0.5 seconds to save the 16k files on Ptycho (cold, in isolation) so I could convert this fake datset into a fixture if you prefer? Would mean we check the file limit on a wider variety of configurations, too.